edb/schema/delta.py

```diff
@@ -990,36 +990,111 @@ def _build_alter_cmd_stack(
         root = DeltaRoot()
         return root, root
 
-    def _prohibit_if_expr_refs(
+    def _propagate_if_expr_refs(
         self,
         schema: s_schema.Schema,
         context: CommandContext,
         action: str,
-    ) -> None:
+    ) -> Tuple[s_schema.Schema, List[qlast.DDLCommand]]:
         scls = self.scls
         expr_refs = s_expr.get_expr_referrers(schema, scls)
+        # Commands to be executed after the original change is
+        # complete
+        finalize_ast: List[qlast.DDLCommand] = []
 
         if expr_refs:
             ref_desc = []
             for ref, fn in expr_refs.items():
-                if fn == 'expr':
-                    fdesc = 'expression'
+
+                from edb.common.markup import dump
+                dump(ref, schema=schema, marker='delta.py:1006')
+
+                from . import ddl as s_ddl
+                from . import indexes as s_indexes
+
+                if isinstance(ref, s_indexes.Index):
+                    # If the affected entity is an index, just drop it and
+                    # schedule it to be re-created.
+                    idx_subj = ref.get_subject(schema)
+
+                    dump(ref.get_expr(schema).origqlast, schema=schema, marker='delta.py:1020')
+                    dump(idx_subj, schema=schema, marker='delta.py:1021')
+                    dump(idx_subj.get_name(schema), schema=schema, marker='delta.py:1022')
+
+                    module, shortname = idx_subj.get_name(schema).split('::', 1)
+
+                    finalize_ast.append(
```

Comment:
See `ReferencedObject.delete` and `_propagate_field_alter` on how to use `_build_alter_cmd_stack` to build arbitrary alteration subtrees.  This should also need to be factored out in a way that allows moving index-specific operations to `indexes.py` etc.